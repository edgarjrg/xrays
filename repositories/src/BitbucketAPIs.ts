import * as Bluebird from "bluebird";
import fetch from "node-fetch";
import { Auth } from "../../auth/src";
import { repositories } from "../../creds/repos";
import {
  ACTIVITY_IN_PULL_REQUESTS_TEMPLATE,
  ACTIVITY_IN_PULL_REQUEST_TEMPLATE,
  AUTHORIZATION_TEMPLATE,
  COMMENTS_IN_PULL_REQUEST_TEMPLATE,
  PULL_REQUESTS_TEMPLATE,
  REPOSITORIES_TEMPLATE,
  PULL_REQUEST_TEMPLATE
} from "./apiTemplates";

function fetchFromAPI<T>({ Authorization, URL }: { Authorization: string, URL: string }): Bluebird<T> {
  return Bluebird
    .resolve(
      fetch(URL, {
        headers: {
          Authorization,
        },
        method: "GET",
      }))
    .then((response) => response.json() as Promise<T>);
}

export class BitbucketAPI {
  public auth: Auth;
  constructor(auth) {
    this.auth = auth;
  }

  public repositories(): Bluebird<IRepository[]> {
    return Bluebird.resolve(repositories);
  }

  public userRepositories(repository: IRepository): Bluebird<any> {
    return this.auth.logIn()
      .then((authentication) =>
        fetchFromAPI({
          Authorization: AUTHORIZATION_TEMPLATE(authentication.access_token),
          URL: REPOSITORIES_TEMPLATE(repository),
        }));
  }

  public activityInPullRequests(repository: IRepository): Bluebird<any> {
    return this.auth.logIn()
      .then((authentication) =>
        fetchFromAPI({
          Authorization: AUTHORIZATION_TEMPLATE(authentication.access_token),
          URL: ACTIVITY_IN_PULL_REQUESTS_TEMPLATE(repository),
        }));
  }

  public commentsInPullRequest(repository: IRepository, pullRequestID: number): Bluebird<CommentsInPullRequest> {
    return this.auth.logIn()
      .then((authentication) =>
        fetchFromAPI<CommentsInPullRequest>({
          Authorization: AUTHORIZATION_TEMPLATE(authentication.access_token),
          URL: COMMENTS_IN_PULL_REQUEST_TEMPLATE(repository, pullRequestID)
        }))
  }

  public retrieveComment(url) {
    return this.auth.logIn()
      .then((authentication) =>
        fetchFromAPI({
          Authorization: AUTHORIZATION_TEMPLATE(authentication.access_token),
          URL: url
        }))
  }

  public activityInPullRequest(repository: IRepository, pullRequestID: number): Bluebird<any> {
    return this.auth.logIn()
      .then((authentication) =>
        fetchFromAPI<any>({
          Authorization: AUTHORIZATION_TEMPLATE(authentication.access_token),
          URL: ACTIVITY_IN_PULL_REQUEST_TEMPLATE(repository, pullRequestID),
        }));
  }

  public pullRequests(repository: IRepository, state?: PULL_REQUEST_STATE, page?: number): Bluebird<any> {
    return this.auth.logIn()
      .then((authentication) =>
        fetchFromAPI<any>({
          Authorization: AUTHORIZATION_TEMPLATE(authentication.access_token),
          URL: PULL_REQUESTS_TEMPLATE(repository, state, page),
        }));
  }

  public pullRequest(repository: IRepository, pullRequestID: number): Bluebird<any> {
    return this.auth.logIn()
      .then((authentication) =>
        fetchFromAPI<any>({
          Authorization: AUTHORIZATION_TEMPLATE(authentication.access_token),
          URL: PULL_REQUEST_TEMPLATE(repository, pullRequestID),
        }));
  }
}