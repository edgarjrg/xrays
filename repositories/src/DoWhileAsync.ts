import * as Bluebird from 'bluebird'

export async function DoWhileAsync<T>(
  fn: () => Bluebird<T>,
  continueFetching: (arg: T) => boolean,
  dataCallback: (arg: T) => void) {

  let response;

  do {
    response = await fn()
    dataCallback(response)
  } while (continueFetching(response))

}