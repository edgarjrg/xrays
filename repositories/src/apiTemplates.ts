import { IRepository } from "./IRepository";

export const api2: string = "https://api.bitbucket.org/2.0/";

// export const api1: string = "https://bitbucket.org/rest/api/1.0";

export const AUTHORIZATION_TEMPLATE =
    (accessToken: string): string =>
        `Bearer ${accessToken}`;

export const COMMENTS_IN_PULL_REQUEST_TEMPLATE =
    ({ userName, repoSlug }: IRepository, pullRequestId: number): string =>
        `${api2}repositories/${userName}/${repoSlug}/pullrequests/${pullRequestId}/comments`;

export const ACTIVITY_IN_PULL_REQUESTS_TEMPLATE =
    ({ userName, repoSlug }: IRepository): string =>
        `${api2}repositories/${userName}/${repoSlug}/pullrequests/activity`;

export const ACTIVITY_IN_PULL_REQUEST_TEMPLATE =
    ({ userName, repoSlug }: IRepository, pullRequestId: number): string =>
        `${api2}repositories/${userName}/${repoSlug}/pullrequests/${pullRequestId}/activity`;

// export function TASKS_IN_PULL_REQUEST_TEMPLATE({ repoSlug, projectKey }: IRepository, pullRequestId: number): string {
//     return `${api1}/projects/${projectKey}/repos/${repoSlug}/pull-requests/${pullRequestId}/tasks`;
// }
