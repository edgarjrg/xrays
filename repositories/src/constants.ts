export enum PULL_REQUEST_STATE {
  MERGED = 'MERGED',
  SUPERSEDED = 'SUPERSEDED',
  OPEN = 'OPEN',
  DECLINED = 'DECLINED',
}