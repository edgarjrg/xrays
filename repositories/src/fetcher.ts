import * as Bluebird from 'bluebird'

export class Fetcher {

  async start<T>(
    fn: () => Bluebird<T>,
    continueFetching: (arg: T) => boolean,
    dataCallback: (arg: T) => void) {

    let response;

    do {
      response = await fn()
      dataCallback(response)
    } while (continueFetching(response))

  }
}