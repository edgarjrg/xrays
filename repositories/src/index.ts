import * as  Promise from "bluebird";
import fetch from "node-fetch";
import { Auth } from "../../auth/src";
import { repositories } from "../../creds/repos";
import {
  ACTIVITY_IN_PULL_REQUESTS_TEMPLATE,
  AUTHORIZATION_TEMPLATE,
  COMMENTS_IN_PULL_REQUEST_TEMPLATE,
} from "./apiTemplates";
import { IRepository } from "./IRepository";

function fetchFromAPI({ Authorization, URL }) {
  return Promise
    .resolve(
      fetch(URL, {
        headers: {
          Authorization,
        },
        method: "GET",
      }))
    .then((response) => response.json());
}

export class Repository {
  public auth: Auth;
  constructor(auth) {
    this.auth = auth;
  }

  public repositories(): Promise<IRepository> {
    return Promise.resolve(repositories);
  }

  public activityInPullRequests(repository) {
    return this.auth.logIn()
      .then((authentication) =>
        fetchFromAPI({
          Authorization: AUTHORIZATION_TEMPLATE(authentication.access_token),
          URL: ACTIVITY_IN_PULL_REQUESTS_TEMPLATE(repository),
        }));
  }

  public commentsInPullRequest(repository, pullRequestID) {
    return this.auth.logIn()
      .then((authentication) =>
        fetchFromAPI({
          Authorization: AUTHORIZATION_TEMPLATE(authentication['access_token']),
          URL: COMMENTS_IN_PULL_REQUEST_TEMPLATE(repository, pullRequestID)
        }))
  }
}

// // manual tests
// const r = new Repository(new Auth());

// r.repositories()
//   .tap(console.log)
//   .then((repositories) => r.commentsInPullRequest(repositories[0], 330))
//   .then(JSON.stringify)
//   .tap(console.log)
//   .catch(console.error);
