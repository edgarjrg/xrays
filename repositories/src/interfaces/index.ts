export * from "./CommentsInPullRequest";

export * from './IRepository'