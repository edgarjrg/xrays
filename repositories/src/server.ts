import * as Koa from 'koa';
import * as koaRouter from 'koa-router';
import * as koaBody from 'koa-bodyparser';
import { graphqlKoa, graphiqlKoa } from 'apollo-server-koa';
import {
  GraphQLSchema,
  GraphQLObjectType,
  GraphQLString,
  GraphQLInt,
  GraphQLList,

} from 'graphql';

import { db } from "../../sequelize";

const Repository = new GraphQLObjectType({
  name: 'Repository',
  fields: {
    id: {
      type: GraphQLInt
    },
    projectKey: {
      type: GraphQLString
    },
    repoSlug: {
      type: GraphQLString
    },
    userName: {
      type: GraphQLString
    },
    createdAt: {
      type: GraphQLString
    },
    updatedAt: {
      type: GraphQLString
    },
  }
})

var schema = new GraphQLSchema({
  query: new GraphQLObjectType({
    name: 'root',
    fields: {
      repositories: {
        type: new GraphQLList(Repository),
        resolve(obj, args, { db }) {
          return db.models.Repository
            .findAll()
        }
      }
    }
  })
});

const app = new Koa();
const router = new koaRouter();
const PORT = 3000;

const graphqlServer = graphqlKoa({
  schema,
  context: {
    db: db()
  }
})


router.post('/graphql', koaBody(), graphqlServer);
router.get('/graphql', graphqlServer);

router.get('/graphiql', graphiqlKoa({ endpointURL: '/graphql' }));

app.use(router.routes());
app.use(router.allowedMethods());
app.listen(PORT, () => {
  console.log(`listening on port ${PORT}`)
});