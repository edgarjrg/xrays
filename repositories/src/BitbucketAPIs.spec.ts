import { Repository } from "./BitbucketAPIs";
import { Auth } from "../../auth/src/index";

describe('pullRequests', () => {
  let repository: Repository
  const repo = {
    projectKey: "atlassian-oauth",
    repoSlug: "atlassian-oauth",
    userName: "atlassian",
  }
  beforeAll(() => {
    repository = new Repository(new Auth())
  })

  test('should retrieve MERGED pull requests', async () => {
    const response = await repository.pullRequests(repo, 'MERGED')
    expect(Object.keys(response)).toMatchSnapshot()
  });

  test('should retrieve DECLINED pull requests', async () => {
    const response = await repository.pullRequests(repo, 'DECLINED')
    expect(Object.keys(response)).toMatchSnapshot()
  });

  test('should retrieve OPEN pull requests', async () => {
    const response = await repository.pullRequests(repo, 'OPEN')
    expect(Object.keys(response)).toMatchSnapshot()
  });

  test('should retrieve SUPERSEDED pull requests', async () => {
    const response = await repository.pullRequests(repo, 'SUPERSEDED')
    expect(Object.keys(response)).toMatchSnapshot()
  });
})

test('should retrieve username projects', async () => {
  const repository = new Repository(new Auth())
  const repo = {
    projectKey: "atlassian-oauth",
    repoSlug: "atlassian-oauth",
    userName: "atlassian",
  }

  const response = await repository.userRepositories(repo)
  expect(Object.keys(response)).toMatchSnapshot()

});

