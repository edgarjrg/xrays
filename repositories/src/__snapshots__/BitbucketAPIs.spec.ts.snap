// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pullRequests should retrieve DECLINED pull requests 1`] = `
Array [
  "pagelen",
  "values",
  "page",
  "size",
]
`;

exports[`pullRequests should retrieve MERGED pull requests 1`] = `
Array [
  "pagelen",
  "size",
  "values",
  "page",
  "next",
]
`;

exports[`pullRequests should retrieve OPEN pull requests 1`] = `
Array [
  "pagelen",
  "values",
  "page",
  "size",
]
`;

exports[`pullRequests should retrieve SUPERSEDED pull requests 1`] = `
Array [
  "pagelen",
  "values",
  "page",
  "size",
]
`;

exports[`should retrieve username projects 1`] = `
Array [
  "next",
  "values",
  "pagelen",
  "size",
  "page",
  "previous",
]
`;
